### DISCUSSION TITLE
Pushing from Codespace on a repo where you aren't a collaborator

### DISCUSSION THREAD

**Discussion by simeydk** [community/community#7795] [Codespaces]
# Cannot push after creating Codespace from upstream

I've run into this problem twice now, and I don't know how to resolve it. I go to a repo that I don't own and want to make a contribution, using codespaces. I click the green code button and create a new codespace. I go on my way coding up my contribution. 

Then, when I want to push my code to create a PR, I get stuck. 

I have tried forking the repo using the `gh` cli, but got an error. 
I then forked the repo on github.com, and added my fork as a remote using `git remote add` in the codespace. This works, but then  when I try `git push` I get an error. 

Here are the steps to reproduce and the errors I got:

### aside
I know I _should have_ forked the repo first and then created the codespace on the fork. However, I didn't, so how do I get my changes out of the codespace given that I'm already here. 

Moreover, if this is always an issue, shouldn't the web ui prevent you from creating a codespaces directly on a repo you don't "own"?

## Steps to reproduce

1. Go to a repo you don't own (e.g. [facebook/react](https://github.com/facebook/react) for me)
2. Create a new codespace directly on the main repo: Click the Green Code button > Codespaces > create new codespace
3. Make a change and commit

Now fork the repo to create a PR

4. Fork the repo using the `gh` cli in the codespace: Run `gh repo fork` and get the following error:
```
$ gh repo fork
failed to fork: HTTP 403: Resource not accessible by integration (https://api.github.com/repos/gatsbyjs/gatsby-starter-minimal/forks)
```
or
5. go to the repo on github, and click fork. Click the Code button, go to the local tab and copy the HTTPS link (https://github.com/{{your_username}}/react.git)
6. in the codespace, run `git remote add myfork https://github.com/{{your_username}}/react.git`
7. in the codespace, push changes to a new branch on your fork on github, and get the following error:
```
$ git push --set-upstream myfork mybranch
remote: Permission to {{your_username}}/react.git denied to {{your_username}}.
fatal: unable to access 'https://github.com/{{your_username}}/react.git/': The requested URL returned error: 403
```

What should I do to be able to push my changes to my fork?

**Comment by bartr** [comment]
When you create a Codespace, a PAT is generated that is scoped to that repo. `echo $GITHUB_TOKEN`

Git is configured to use GITHUB_TOKEN for auth `cat /.codespaces/bin/gitcredential_github.sh`

export GITHUB_TOKEN=YourRealPAT and everything works.

I have a Codespaces secret (PAT) and if PAT is set, I add `export GITHUB_TOKEN=$PAT` to my .zshrc via my dotfiles.

**Reply by xiangqunXiao** [reply]
I do not manual set PAT, and `echo $PAT` returns nothing. So what should I do to solve it?

cat /.codespaces/bin/gitcredential_github.sh :
```
#!/bin/sh
echo protocol=https
echo host=github.com
echo path=
echo username=PersonalAccessToken
echo password=$GITHUB_TOKEN

```

**Reply by Christopher-Hayes** [reply]
Worked for me! Thanks @bartr!

@xiangqunXiao you gotta create a token through GitHub Developer settings, and give it access to all repos.

GitHub developer settings: 
![image](https://user-images.githubusercontent.com/6013871/170516104-eab04331-2d3d-40e8-95ed-df833675b315.png)

If you then run `export GITHUB_TOKEN=<PASTE YOUR TOKEN HERE>` (For example: `export GITHUB_TOKEN=gabisfhowiehgonoiewhoafjoai`). The $PAT thing is in reference to the fact that Codespaces lets you make "secrets" that are accessible from your dev containers.

To set up `$PAT` go to you Codespaces settings: https://github.com/settings/codespaces
![image](https://user-images.githubusercontent.com/6013871/170516687-b1b3418c-c373-466a-bbc8-c4ff915312e4.png)


Create a secret called "PAT" with the token you created above.
![image](https://user-images.githubusercontent.com/6013871/170516843-cb03c33a-cb36-43e5-b024-8600daf32505.png)

Now it's a little easier, since you just need to run `export GITHUB_TOKEN=$PAT`. However, note that the secret only is accessible on the repos you set. So, each time you use a new repo, you still have to go to your "Codespaces Secrets" settings and add the new repo.

OP mentions using a dotfile. Codespaces let's you connect use your own personal dotfiles to your codespace. If you add `export GITHUB_TOKEN=$PAT` to your Codespace-connected dotfile, then you don't have to run that command anymore.
![image](https://user-images.githubusercontent.com/6013871/170518118-04c4aa5c-06be-4209-bb52-e53455114513.png)

**Reply by bartr** [reply]
Much deeper explanation!

In my .zshrc, I have an if block that checks "$PAT" != ""

As Christopher points out, if the repo isn't authorized for your Codespaces secret named PAT, it is blank and without the if block, you overwrite GITHUB_TOKEN with "" and can't push to any repo.

There are several other uses for this approach
- we create ID_RSA and ID_RSA_PUB secrets (org, repo, or personal) and save them to ~/.ssh/id_rsa[.pub]
  - cat id_rsa | base64 | tr -d '\n' to get the value for the secret
  - cat $ID_RSA | base64 -d > ~/.ssh/id_rsa  to save the value
 - I have an Azure SP that I login with (WAY better than az login --use-device-code)
  - TENANT, SP_ID, SP_KEY
  - This let's me use Azure Key Vault very easily
- we have our SSL pem and key as secrets

Really any key / config value you want to save.

If you prefix all of your secrets (i.e. CSSEC_), you can run `env | grep CSSEC_` and see what's set and not. I takes a little getting used to the behavior, but once you do, it's VERY powerful.

I work for Microsoft, but not GitHub - I'm just a big fan of Codespaces :) Use it every day.

**Comment by Christopher-Hayes** [comment]
GitHub previously allowed you to configure your access token not to be restricted, but they deprecated it at some point. I agree though, you should be able to push your changes to your fork.
![image](https://user-images.githubusercontent.com/6013871/170513987-a04da5f9-cdf7-4733-b41f-945c8c88c32e.png)
